wandb/sdk/wandb_run.py

```diff
@@ -184,31 +184,36 @@ def join(self) -> None:
         self._retry_thread.join()
 
 
-def _attach(func: Callable) -> Callable:
-    @functools.wraps(func)
-    def wrapper(self: Type["Run"], *args: Any, **kwargs: Any) -> Any:
-
-        # * `_attach_id` is only assigned in service hence for all non-service cases
-        # it will be a passthrough.
-        # * `_init_pid` is only assigned in __init__:
-        #   - for non-fork case the object is shared through pickling so will be None.
-        #   - for fork case the new process share mem space hence the value would be of parent process.
-        if (
-            getattr(self, "_attach_id", None)
-            and getattr(self, "_init_pid", None) != os.getpid()
-        ):
-            if self._is_attaching:
-                message = f"Trying to attach `{func.__name__}` while in the middle of attaching `{self._is_attaching}`"
-                raise RuntimeError(message)
-            self._is_attaching = func.__name__
-            try:
-                wandb._attach(run=self)
-            except Exception as e:
-                raise e
-        self._is_attaching = ""
-        return func(self, *args, **kwargs)
-
-    return wrapper
+class Attach:
+
+    _is_attaching: str = ""
+
+    @classmethod
+    def _attach(cls, func: Callable) -> Callable:
+        @functools.wraps(func)
+        def wrapper(self: Type["Run"], *args: Any, **kwargs: Any) -> Any:
+
+            # * `_attach_id` is only assigned in service hence for all non-service cases
+            # it will be a passthrough.
+            # * `_attach_pid` is only assigned in _init:
+            #   - for non-fork case the object is shared through pickling so will be None.
+            #   - for fork case the new process share mem space hence the value would be of parent process.
+            if (
+                getattr(self, "_attach_id", None)
+                and getattr(self, "_attach_pid", None) != os.getpid()
+            ):
+                if cls._is_attaching:
+                    message = f"Trying to attach `{func.__name__}` while in the middle of attaching `{cls._is_attaching}`"
+                    raise RuntimeError(message)
+                cls._is_attaching = func.__name__
+                try:
+                    wandb._attach(run=self)
+                except Exception as e:
+                    raise e
```

Comment:
No need to reset `cls._is_attaching` on error?

Reply:
hmmm... yeah I can add it to be safe... just assumed if this happens we just crush the program...

Reply:
yea, I would also add an informative comment before raising the error.

Reply:
what do you mean by informative comment?